src/GoogleFitOnFhir/Services/GoogleFitImportService.cs

```diff
@@ -0,0 +1,119 @@
+
+        public GoogleFitImportService(
+            IGoogleFitClient googleFitClient,
+            EventHubProducerClient eventHubProducerClient,
+            GoogleFitImportOptions options,
+            ILogger<GoogleFitImportService> logger,
+            ITelemetryLogger telemetryLogger)
+            : base(options, options?.ParallelTaskOptions?.MaxConcurrency ?? 1)
+        {
+            _googleFitClient = EnsureArg.IsNotNull(googleFitClient, nameof(googleFitClient));
+            _eventHubProducerClient = EnsureArg.IsNotNull(eventHubProducerClient, nameof(eventHubProducerClient));
+            _logger = EnsureArg.IsNotNull(logger, nameof(logger));
+            _telemetryLogger = EnsureArg.IsNotNull(telemetryLogger, nameof(telemetryLogger));
+        }
+
+        public async Task ProcessDatasetRequests(
+            User user,
+            IEnumerable<DataSource> dataSources,
+            string datasetId,
+            AuthTokensResponse tokensResponse,
+            CancellationToken cancellationToken)
+        {
+            var workItems = dataSources.Select(
+                dataSource => new Func<Task>(
+                async () =>
+                {
+                    var dataStreamId = dataSource.DataStreamId;
+                    string pageToken = null;
+
+                    try
+                    {
+                        // Make the Dataset requests, requesting the next page of data if necessary
+                        do
+                        {
+                            _logger.LogInformation("Query Dataset: {0}", dataStreamId);
+                            var medTechDataset = await _googleFitClient.DatasetRequest(
+                                tokensResponse.AccessToken,
+                                dataSource,
+                                datasetId,
+                                cancellationToken,
+                                pageToken);
+
+                            if (medTechDataset == null)
+                            {
+                                _logger.LogInformation("No Dataset for: {0}", dataStreamId);
+                                continue;
+                            }
+
+                            // Save the NextPageToken
+                            pageToken = medTechDataset.GetDataset().NextPageToken;
+
+                            _logger.LogInformation("Create Eventhub Batch");
+
+                            // Create a batch of events for IoMT eventhub
+                            using var eventBatch = await _eventHubProducerClient.CreateBatchAsync(cancellationToken);
+
+                            _logger.LogInformation("Push Dataset: {0}", dataStreamId);
+
+                            // Push dataset to IoMT connector
+                            if (!eventBatch.TryAdd(medTechDataset.ToEventData(user.Id)))
+                            {
+                                throw new Exception("Event is too large for the batch and cannot be sent.");
```

Comment:
I think we should create a new exception and log t as an error, but not throw. What I have seen happen in this situation is that the do loop is broken and the next page(s) are not reteived, so we orphan data. We want to know that the condition occured, but we don't want the condition to prevent more data from being processed.